### PR TITLE
BBS-171 - implement filters for primo

### PR DESCRIPTION
**Pull request by danquah**

### CODE REVIEW

File: modules/primo/src/Primo/BriefSearch/Client.php [hunk 1]
```diff
@@ -165,4 +124,47 @@ public function validateThumbnail($url) {
...
+   */
+  public function search(array $queryParameters, $offset, $numResults) {
+    // Set user-specified "runtime" values.
+    $queryParameters['indx'] = [$offset];
+    $queryParameters['bulkSize'] = [$numResults];
+
+    // Add in "static" defaults.
+    array_walk($this->defaultParameters, function ($param_value, $param_key) use (&$queryParameters) {
+      $queryParameters[$param_key][] = $param_value;
+    });
```
Comment (kasperg) [inline modules/primo/src/Primo/BriefSearch/Client.php]: Is this actually [`array_merge_recursive`](http://php.net/manual/en/function.array-merge-recursive.php)?

File: modules/primo/src/Primo/Ting/PrimoStatementRenderer.php [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+          // "name=value AND (name=value AND name=value)"
+          // equals
+          // "name=value AND name=value AND name=value".
+          $carry = array_merge($carry, array_map(function ($statement) {
+            return new BooleanStatementGroup([$statement]);
+          }, $group_statements));
+        }
+        else {
+          // The group referenced more than one field and was not joined by AND.
+          throw new UnsupportedSearchQueryException("Primo only supports AND between different field.");
```
Comment (kasperg) [inline modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: Nitpicking field**s**.

Also: Would it be possible to provide more information about what fields conflicted or query was attempted?

File: modules/primo/src/Primo/Ting/PrimoStatementRenderer.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  public function escapeValue($value) {
+    $replacements = [
+      // AND and OR are keywords, strip them.
+      ' AND ' => ' ',
+      ' OR ' => ' ',
+      // Commas should be replaced with spaces.
+      ',' => ' ',
+    ];
+
+    return str_replace(array_keys($replacements), array_values($replacements), $value);
```
Comment (kasperg) [inline modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: If we already have a map from search to replace then we can use that with [`strtr`](http://php.net/manual/en/function.strtr.php).

File: modules/primo/src/Primo/Ting/PrimoStatementRenderer.php [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  /**
+   * Renders a list of statements into primo briefsearch queries.
+   *
+   * @param array $statements
+   *   List of statements.
+   *
+   * @return array
+   *   The rendered statements as a list of <key> => query where key is a valid
```
Comment (kasperg) [inline modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: Could we be more specific with the type hints in the DocBlock here?

File: modules/primo/src/Primo/Ting/PrimoStatementRenderer.php [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    // Ensure we can support the statements and normalize the statements into
+    // a list of groups.
+    $groups = $this->preprocessStatements($statements);
+
+    // We now have a list of groups containing one or more fields each. If the
+    // group contains several statements, they will be against the same field.
+    // We can now map the array using $this->renderGroup();
+    return array_reduce(array_map([$this, 'renderGroup'], $groups), function ($carry, $rendered_group) {
+      foreach ($rendered_group as $parameter_key => $parameter_value) {
+        if(isset($carry[$parameter_key])) {
```
Comment (kasperg) [inline modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: Missing ` ` after if here.

File: modules/primo/src/Primo/Ting/PrimoStatementRenderer.php [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+      return $carry;
+    }, []);
+  }
+
+  /**
+   * Verify we can process the statements and convert them to groups.
+   *
+   * @param mixed[] $statements
+   *   One or more instances of BooleanStatementGroup and/or
+   *   TingSearchFieldFilter.
```
Comment (kasperg) [inline modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: This rubs be a bit the wrong way. I wonder if we could introduce a shared interface between the two.

**Review by kasperg — CHANGES_REQUESTED** [review]
Another day, another review 😅

File: modules/primo/src/Primo/Ting/PrimoStatementRenderer.php (hunk 6 above)
Comment (danquah) [reply modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: The inteface would look something like this (unless you're thinking a tag-interface?)
```
interface BooleanStatementMember {
   /**
   * ..
   * @return TingSearchFieldFilter[]
   */
}
```
What would that interface help me achieve? I still have to do an instanceof when handling the actual members of the array (as I want to treat groups differently from fields) so the interface is not helping me signal what you are allowed to pass in to the function as we only support known implementations of the interface.

I get that an "untyped" array is messy - but in this case I'm pretty clear about what I will allow in.

Comment (danquah) [reply modules/primo/src/Primo/Ting/PrimoStatementRenderer.php]: Gave it another look, I'll implement a tag interface, but the bulk of the work has to happen in bbs-sal so I'm putting this PR on hold until I get that fixed and have bbs-master updated.

**Review by kasperg — APPROVED** [review]

**Comment by kasperg** [comment]
:+1: LGTM

**Comment by kasperg** [comment]
Notice that tests fail for some reason.

**Pull request merged**